File: LeetCode/545.Boundary_of_Binary_Tree/Mario/545.Boundary_of_Binary_Tree.py
#!/usr/bin/env python3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def boundaryOfBinaryTree(self, root: Optional[TreeNode]) -> List[int]:
        # [root] + [left boundary] + [leaves] + [right boundary]

        res = []
        if root:
            res.append(root.val)

        def left_bounadry(node):
            if not node:
                return
            if not node.left and not node.right:
                return
            else:
                res.append(node.val)
                if node.left:
                    left_bounadry(node.left)
                elif node.right:
                    left_bounadry(node.right)

        right = []

        def right_bounadry(node):
            if not node:
                return
            if not node.left and not node.right:
                return
            else:
                right.append(node.val)
                if node.right:
                    right_bounadry(node.right)
                elif node.left:
                    right_bounadry(node.left)

        def pre_travseral(node):
            if not node:
                return

            if not node.left and not node.right and node != root:
                res.append(node.val)
                return

            left = pre_travseral(node.left)
            right = pre_travseral(node.right)

        if root.left:
            left_bounadry(root.left)

        pre_travseral(root)
        if root.right:
            right_bounadry(root.right)

        # print(res)
        # print(right)

        return res + right[::-1]

File: LeetCode/545.Boundary_of_Binary_Tree/545.Boundary_of_Binary_Tree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func boundaryOfBinaryTree(_ root: TreeNode?) -> [Int] {
        // [root] + [left boundary] + [leaf] + [right boundary]
        guard let root = root else { return [] }
        var res: [Int] = [root.val]

        func isLeaf(_ node: TreeNode) -> Bool {
            return node.left == nil && node.right == nil
        }

        if isLeaf(root) { return res }

        // left boundary
        func getLeftBoundary(_ node: TreeNode?) {
            guard let node = node else { return }
            var cur: TreeNode? = node

            while let n = cur {
                if isLeaf(n) { break }
                res.append(n.val)
                cur = n.left ?? n.right
            }
        }

        func getRightBoundary(_ node: TreeNode?) {
            guard let node = node else { return }
            var cur: TreeNode? = node
            var ans: [Int] = []

            while let n = cur {
                if isLeaf(n) { break }
                ans.append(n.val)
                cur = n.right ?? n.left
            }

            res.append(contentsOf: ans.reversed())
        }

        func getLeaves(_ root: TreeNode?) {
            guard let root = root else { return }

            if isLeaf(root) { res.append(root.val) }
            getLeaves(root.left)
            getLeaves(root.right)
        }

        getLeftBoundary(root.left)
        getLeaves(root)
        getRightBoundary(root.right)

        return res
    }
}
